### PR TITLE
Issue/10/tomo bins

### LINKED ISSUE #10: Simple class for tomographic binning
I would like to add a very simple class for tomographic binning. It will be using the point estimate, e.g. mode of the photo-z posterior, to split the sample into tomographic bins according to the Science Requirement Document. The class will probably be placed in summarization, but open to suggestions. Comments, advice, and helps are welcome!

### PR DESCRIPTION
**Pull request by hangqianjun**
## Change Description

- [x] My PR includes a link to the issue that I am addressing
Resolves #10 

## Solution Description

In this PR I have added a `tomographer` module under `src/rail/estimation`. It contains two generic tomographer classes:

- `PZTomographer`, which takes per-galaxy n(z) from a `qp.Ensemble` object and output a tabular object with tomographic binning;
- `CatTomographer`, which takes catalogue-like data and output a tabular object with tomographic binning;
The second type will be compatible with the classifiers in tomo-challenge, where features in the catalogue are used to assign tomographic bins.

For each of these types, I've added an example classifier in `algos/`. `naiveClassifierSRD` is a PZtomographer that uses simple point estimate SRD binning; `randomForestClassifier` is a CatTomographer which is adapted from TXPipe.

## Code Quality
- [x] I have read the Contribution Guide
- [x] My code follows the code style of this project
- [ ] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

### Bug Fix Checklist
- [ ] My fix includes a new test that breaks as a result of the bug (if possible)
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### New Feature Checklist
- [x] I have added or updated the docstrings associated with my feature using the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover my new feature
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### Documentation Change Checklist
- [ ] Any updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)

### Build/CI Change Checklist
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the README to reflect this
- [ ] If this is a new CI setup, I have added the associated badge to the README

### Other Change Checklist
- [ ] Any new or updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover any changes
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### CODE REVIEW

**Comment by hangqianjun** [comment]
While trying to writing a unit test for the `PZTomographer` algorithm, I realised that there doesn't seem to be test qp.Ensemble data available in `src/rail/examples_data/testdata/`. Could we add a small test qp file like `test_dc2_training_9816.hdf5`?

**Comment by eacharles** [comment]
Depends on the size.   If it is more that a few 10’s of galaxies it would be better to download the data than to include it in the repo.On Jul 4, 2023, at 12:20 PM, hangqianjun ***@***.***> wrote:﻿
While trying to writing a unit test for the PZTomographer algorithm, I realised that there doesn't seem to be test qp.Ensemble data available in src/rail/examples_data/testdata/. Could we add a small test qp file like test_dc2_training_9816.hdf5?

**Comment by hangqianjun** [comment]
> Depends on the size. If it is more that a few 10’s of galaxies it would be better to download the data than to include it in the repo.
Are these data available somewhere already?

**Comment by eacharles** [comment]
Does `src/rail/examples_data/testdata/output_BPZ_lite.fits` work?

**Comment by sschmidt23** [comment]
@hangqianjun I was looking at this PR earlier, I don't think that we want to have input parameters in a .ini file unless there is a strong reason to do so, as that file could change between runs if someone pushes a change and updated values could lead to non-reproducibility on subsequent runs. Having all of the parameters as config params seems like a better way of tracking things in terms of reproducibility.  Was there a reason to do this with a `.ini` file, e.g. maybe TXPipe does something like that?

Also, `output_BPZ_lite.fits` may not be the best test file, as the mode values for the first 10 galaxies are all either at z<0.2 or z>2.8, and thus the default tomographer value (for SRD binning) for all 10 is `-99`.  We may need to include another small test file here with more appropriate mode values.

**Review by eacharles — APPROVED** [review]
Looks great, thanks Qianjun!

**Comment by hangqianjun** [comment]
Hi @aimalz, are you happy with the module/algorithm names? Let me know if you have suggestions!

File: tests/estimation/test_tomographer.py
Comment (aimalz) [inline tests/estimation/test_tomographer.py:1]: This is a solid coverage test, but I think there are a couple potentially problematic edge cases that we ought to cover in real unit tests (not necessarily in this PR though).

- When `nbins` >1, check that the range of each bin is actually the same in the case of equal bin size.
- When `nbins` >1, check that the number of galaxies per bin is indeed the same in each bin.
- With config parameter values chosen such that nonclassified galaxies are anticipated, check that they end up getting that nonclassified value.

File: src/rail/estimation/algos/naiveClassifierSRD.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+from rail.estimation.tomographer import PZTomographer
+from rail.core.data import TableHandle
+
+class naiveClassifierSRD(PZTomographer):
+    """Classifier that simply assign tomographic
+    bins based on point estimate according to SRD"""
+
+    name = 'naiveClassifierSRD'
+    config_options = PZTomographer.config_options.copy()
+    config_options.update(
```
Comment (aimalz) [inline src/rail/estimation/algos/naiveClassifierSRD.py:17]: Can the value for nonclassified galaxies be a config option specified by the user? I can imagine a case where a user's downstream pipeline might want it to be something other than `-99`.

File: src/rail/estimation/tomographer.py
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/tomographer.py:1]: The name of the class is changed. Currently `CatInformer` is inherited from `estimation.estimator` but there is another issue for isolating the estimators.

File: src/rail/estimation/algos/naiveClassifierSRD.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                cum=np.arange(1,(len(zb)+1))
+                bin_index = np.zeros(len(zb))
+                for ii in range(self.config.nbins):
+                    perc1=ii/self.config.nbins
+                    perc2=(ii+1)/self.config.nbins
+                    ind=(cum/cum[-1]>perc1)&(cum/cum[-1]<=perc2)
+                    useind=sortind[ind]
+                    bin_index[useind] = int(ii+1)
+
+        tomo = {"tomo": bin_index}
```
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/algos/naiveClassifierSRD.py:68]: Changed the key to `class_id`.

File: src/rail/estimation/algos/naiveClassifierSRD.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+"""
+
+import numpy as np
+from ceci.config import StageParameter as Param
+from rail.estimation.tomographer import PZTomographer
+from rail.core.data import TableHandle
+
+class naiveClassifierSRD(PZTomographer):
+    """Classifier that simply assign tomographic
+    bins based on point estimate according to SRD"""
```
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/algos/naiveClassifierSRD.py:13]: Done.

File: src/rail/estimation/algos/naiveClassifierSRD.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+A classifier that uses pz point estimate to assign
+tomographic bins according to SRD
+"""
+
+import numpy as np
+from ceci.config import StageParameter as Param
+from rail.estimation.tomographer import PZTomographer
+from rail.core.data import TableHandle
+
+class naiveClassifierSRD(PZTomographer):
```
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/algos/naiveClassifierSRD.py:11]: This is split into `UniformBinningClassifier` and `EqualCountClassifer`.

File: src/rail/estimation/algos/naiveClassifierSRD.py (hunk 1 above)
Comment (hangqianjun) [reply src/rail/estimation/algos/naiveClassifierSRD.py:17]: Done.

File: src/rail/estimation/tomographer.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        Finally, this will return a TableHandle providing access to that output data.
+
+        Parameters
+        ----------
+        input_data : `dict`
+            A dictionary of all input data
+
+        Returns
+        -------
+        output: `dict`
```
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/tomographer.py:87]: Should we produce catalog_ID + class_ID, or attach the class_ID to the original data?

File: src/rail/estimation/tomographer.py
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/tomographer.py:1]: Now using `estimation.informer`.

File: src/rail/estimation/tomographer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hangqianjun) [reply src/rail/estimation/tomographer.py:87]: Now passing catalog_ID as well as class_ID.

File: tests/estimation/test_tomographer.py
Comment (hangqianjun) [reply tests/estimation/test_tomographer.py:1]: Included in the tests `test_UniformBinningClassifier_binsize`, `test_EqualCountClassifier_nobj`. No assignment cases are tested.

**Review by aimalz — APPROVED** [review]
:100:
